### PR TITLE
Check overlong when decoding UTF-8

### PR DESCRIPTION
**Pull request by geraint0923**
Fix [prestodb/presto#4503](https://github.com/prestodb/presto/issues/4503)

### CODE REVIEW

File: src/test/java/io/airlift/slice/TestSliceUtf8.java [hunk 1]
```diff
@@ -116,6 +116,13 @@
         // min and max surrogate characters
         invalidSequences.add(new byte[] {(byte) 0b11101101, (byte) 0xA0, (byte) 0x80});
         invalidSequences.add(new byte[] {(byte) 0b11101101, (byte) 0xBF, (byte) 0xBF});
+
+        // check overlong encoding
+        invalidSequences.add(new byte[] {(byte) 0b11000000, (byte) 0xAF});
```
Comment (haozhun) [inline src/test/java/io/airlift/slice/TestSliceUtf8.java]: I would add a comment at the end of this line:
`// 2-byte encoding of /`
Additionally, I would use something like `A` instead of `/` so that the comment I just added is easier to read.

Comment (haozhun) [inline src/test/java/io/airlift/slice/TestSliceUtf8.java]: Add another 2-byte test case that tests `7F`, which is the boundary case.

File: src/test/java/io/airlift/slice/TestSliceUtf8.java [hunk 2]
```diff
@@ -116,6 +116,13 @@
...
+        invalidSequences.add(new byte[] {(byte) 0b11100000, (byte) 0x90, (byte) 0x80});
```
Comment (haozhun) [inline src/test/java/io/airlift/slice/TestSliceUtf8.java]: I would add a comment that says `// 3-byte encoding of 0x100`, and then also add a test case that tests 7FF.

File: src/test/java/io/airlift/slice/TestSliceUtf8.java [hunk 3]
```diff
@@ -116,6 +116,13 @@
...
+        invalidSequences.add(new byte[] {(byte) 0b11110000, (byte) 0x8D, (byte) 0xA0, (byte) 0x80});
```
Comment (haozhun) [inline src/test/java/io/airlift/slice/TestSliceUtf8.java]: Please also add a similar comment.

File: src/test/java/io/airlift/slice/TestSliceUtf8.java [hunk 4]
```diff
@@ -116,6 +116,13 @@
...
+        invalidSequences.add(new byte[] {(byte) 0b11110000, (byte) 0x8D, (byte) 0xBF, (byte) 0xBF});
```
Comment (haozhun) [inline src/test/java/io/airlift/slice/TestSliceUtf8.java]: Please also add a similar comment.

Comment (haozhun) [inline src/test/java/io/airlift/slice/TestSliceUtf8.java]: Finally, add a test case for 0xFFFF.

File: src/main/java/io/airlift/slice/SliceUtf8.java [hunk 5]
```diff
@@ -466,8 +466,9 @@ public static int tryGetCodePointAt(Slice utf8, int position)
 
         if (length == 2) {
             // 110x_xxxx 10xx_xxxx
-            return ((firstByte & 0b0001_1111) << 6) |
+            int codePoint = ((firstByte & 0b0001_1111) << 6) |
                     (secondByte & 0b0011_1111);
+            return codePoint < 0x80 ? -2 : codePoint;
```
Comment (haozhun) [inline src/main/java/io/airlift/slice/SliceUtf8.java]: Add a comment before this line that mentions overlong encoding. It's not super clear to readers why we are doing this otherwise.

File: src/main/java/io/airlift/slice/SliceUtf8.java [hunk 6]
```diff
@@ -491,7 +492,7 @@ public static int tryGetCodePointAt(Slice utf8, int position)
             if (MIN_SURROGATE <= codePoint && codePoint <= MAX_SURROGATE) {
                 return -3;
             }
-            return codePoint;
+            return codePoint < 0x800 ? -3 : codePoint;
```
Comment (haozhun) [inline src/main/java/io/airlift/slice/SliceUtf8.java]: Do the same.

File: src/main/java/io/airlift/slice/SliceUtf8.java [hunk 7]
```diff
@@ -512,7 +513,7 @@ public static int tryGetCodePointAt(Slice utf8, int position)
                     ((thirdByte & 0b0011_1111) << 6) |
                     (forthByte & 0b0011_1111);
             // 4 byte code points have a limited valid range
-            if (codePoint < 0x11_0000) {
+            if (codePoint < 0x11_0000 && codePoint >= 0x1_0000) {
```
Comment (haozhun) [inline src/main/java/io/airlift/slice/SliceUtf8.java]: And here as well.

File: src/test/java/io/airlift/slice/TestSliceUtf8.java [hunk 8]
```diff
@@ -116,6 +116,17 @@
...
         invalidSequences.add(new byte[] {(byte) 0b11101101, (byte) 0xBF, (byte) 0xBF});
+
+        // check overlong encoding
+        invalidSequences.add(new byte[] {(byte) 0b11000000, (byte) 0xAF}); // 2-byte encoding of 0x2F
+        invalidSequences.add(new byte[] {(byte) 0b11000001, (byte) 0xBF}); // 2-byte encoding of 0x7F
+        invalidSequences.add(new byte[] {(byte) 0b11100000, (byte) 0x81, (byte) 0xBF}); // 3-byte encoding of 0x7F
+        invalidSequences.add(new byte[] {(byte) 0b11100000, (byte) 0x90, (byte) 0x80}); // 3-byte encoding of 0x400
+        invalidSequences.add(new byte[] {(byte) 0b11100000, (byte) 0x9F, (byte) 0xBF}); // 3-byte encoding of 0x7FF
+        invalidSequences.add(new byte[] {(byte) 0b11110000, (byte) 0x8D, (byte) 0xA0, (byte) 0x80}); // 4-byte encoding of 0xD800
+        invalidSequences.add(new byte[] {(byte) 0b11110000, (byte) 0x8D, (byte) 0xBF, (byte) 0xBF}); // 4-byte encoding of 0xDBEF
```
Comment (haozhun) [inline src/test/java/io/airlift/slice/TestSliceUtf8.java]: the comment is incorrect

File: src/main/java/io/airlift/slice/SliceUtf8.java [hunk 9]
```diff
@@ -511,8 +514,8 @@ public static int tryGetCodePointAt(Slice utf8, int position)
                     ((secondByte & 0b0011_1111) << 12) |
                     ((thirdByte & 0b0011_1111) << 6) |
                     (forthByte & 0b0011_1111);
-            // 4 byte code points have a limited valid range
-            if (codePoint < 0x11_0000) {
+            // 4 byte code points have a limited valid range and check overlong encoding
```
Comment (haozhun) [inline src/main/java/io/airlift/slice/SliceUtf8.java]: Check for overlong encoding and code points above upper bound of Unicode.

**Comment by haozhun** [comment]
Merged #52
